fempy/FitCF.py

```diff
@@ -1,26 +1,242 @@
 
+Usage:
+python3 CorrelationFitter.py cfg.yml
+
+'''
+
+import os
+import argparse
+import yaml
+import ctypes
+
+from ROOT import TFile, TCanvas, gInterpreter, TH1, TH1D, TSpline3
+
+from fempy import logger as log
 from fempy.utils.io import Load
+from fempy.utils.analysis import ChangeUnits
+
+parser = argparse.ArgumentParser()
+parser.add_argument('cfg', default='')
+parser.add_argument('--debug', default=False, action='store_true')
+parser.add_argument('--debugfit', default=False, action='store_true')
+parser.add_argument('--debugdraw', default=False, action='store_true')
+args = parser.parse_args()
+
+if args.debugfit:
+    log.setLevel(1)
+
+if args.debugfit:
+    gInterpreter.ProcessLine(f'#define LOG_LEVEL_FIT 1')
+    
+if args.debugdraw:
+    gInterpreter.ProcessLine(f'#define LOG_LEVEL_DRAW 1')
+    
+gInterpreter.ProcessLine(f'#include "{os.environ.get("FEMPY")}fempy/CorrelationFitter.hxx"')
+gInterpreter.ProcessLine(f'#include "{os.environ.get("FEMPY")}fempy/DrawFitFuncts.hxx"')
+from ROOT import CorrelationFitter, DrawFitFuncts
+
+# Load yaml file
+with open(args.cfg, "r") as stream:
+    try:
+        cfg = yaml.safe_load(stream)
+    except yaml.YAMLError as exc:
+        log.critical('Yaml configuration could not be loaded. Is it properly formatted?')
+
+# Load input file with data and mc CF
+inFileFit = TFile(cfg['infile'])
+
+# Define the output file
+oFileBaseName = cfg['ofilebasename']
+if cfg['suffix'] != '' and cfg['suffix'] is not None:
+    oFileBaseName += f'_{cfg["suffix"]}'
+oFileName = os.path.join(cfg['odir'], oFileBaseName + '.root')
+
+# Open the output file
+try:
+    oFile = TFile.Open(oFileName, 'recreate')
+except OSError:
+    log.critical('The output file %s is not writable', oFileName)
+
+modelFitters = []
+drawFits = []
+
+# for loop over the correlation functions
+for iFit, fitcf in enumerate(cfg['fitcfs']):
+    
+    # change unity of measure of histograms from GeV to MeV
+    fitHisto = ChangeUnits(Load(inFileFit, fitcf['cfpath']), 1000)
+    if fitcf.get('rebin'):
+        fitHisto.Rebin(fitcf['rebin'])
+
+    # fit range
+    lowFitRange = fitcf['fitrange'][0]
+    uppFitRange = fitcf['fitrange'][1]
+    if fitcf.get('rejectrange'):
+        lowRejectRange = fitcf['rejectrange'][0]
+        uppRejectRange = fitcf['rejectrange'][1]
+        modelFitters.append(CorrelationFitter(fitHisto, lowFitRange, uppFitRange, lowRejectRange, uppRejectRange))
+    else: 
+        modelFitters.append(CorrelationFitter(fitHisto, lowFitRange, uppFitRange))
+        
+    # drawing class constructor
+    if fitcf.get('drawrange'):
+        lowDrawRange = fitcf['drawrange'][0]
+        uppDrawRange = fitcf['drawrange'][1]
+        drawFits.append(DrawFitFuncts(fitHisto, lowDrawRange, uppDrawRange))
+    else: 
+        drawFits.append(DrawFitFuncts(fitHisto, lowFitRange, uppFitRange))
+
+    # directory of the fit
+    oFile.mkdir(fitcf['fitname'])
+    oFile.cd(fitcf['fitname'])
+
```

Comment:
I was using the custom name of the folders containing the fit results to gather several tries with different fit functions on the same histogram in one file. The naming of the histos in the folder containing the fit result are standard. I would prefer keeping it like this, also because I think that the .root file containing the fit is the end of the chain for the analysis so it will hardly become an input file for a further step in the analysis.